CentOS-PaaS-SIG/linch-pin
# Linchpin hooks :: resolves #96

Summary:
- linchpin now supports hooks in Pinfile in following states 
  preup, predown, postup and post down 
  ```
  ref: example below
  ```
- hooks are resolved by name ie., if we have a hook named do_something of ansible type the workspace should contain following layout with all the playbooks.
```
.
├── hooks
│   ├── ansible
│   │   ├── do_something
│   │   │   ├── test_playbook.yaml
│   │   │   ├── test_vars.yaml
```
example Pinfile for above ansible hook looks as follows:
```
---
openstack:
  topology: ex_os_server.yml
  layout: openshift-3node-cluster.yml
  hooks:
    preup:              
      - name: postdown_task
        type: ansible
        actions:
          - playbook: test_playbook.yaml
            vars: test_var.yaml
            extra_vars: { "test_var": "postdowntask"}
```
- hooks also supports custom action manager , where the action managers need to implement the following linchpin.hooks.action_manager.ActionManager abstract class .
for example if user would like to implement their own action_manager of type testtype :
hooks folder layout looks as follows : 
```
.
├── hooks
│   └── testtype
│       └── test_something
│           ├── testact.py
│           └── testact.pyc

```
example Pinfile for the above testtype hook looks as follows:
```
openstack:
  topology: ex_os_server.yml
  layout: openshift-3node-cluster.yml
  hooks:
    preup:              
      - name: test_something
        action_manager: testact.py
        type: testtype
        actions:
          - "This does nothing"
```

Example Pinfile:
```
---
openstack:
  topology: ex_os_server.yml
  layout: openshift-3node-cluster.yml
  hooks:
    postup:            
        - name: do_something
          type: python
          context: false
          actions:
            - test.py
        - name: testruby
          type: ruby
          context: false
          actions:
            - test.rb
        - name: testnode
          type: node
          context: true
          actions:
            - test.js
    preup:              
        - name: test_something
          action_manager: testact.py
          type: testtype
          actions:
            - "This does nothing"
        - name: post_up
          type: ansible
          context: false
          actions:
            - playbook: test_playbook.yaml
              vars: test_var.yaml
              extra_vars: { "test_var": "preup task"}
    postdown:
        - name: do_something
          type: shell
          actions:
            - echo " this is post down operation Hello hai how r u ?"
        - name: postdown_task
          type: ansible
          actions:
            - playbook: test_playbook.yaml
              vars: test_var.yaml
              extra_vars: { "test_var": "postdowntask"}
```
current workspace layout required for testing the above Pinfile
```
.
├── hooks
│   ├── ansible
│   │   ├── build_openshift_cluster
│   │   │   ├── test_playbook.yaml
│   │   │   └── test_var.yaml
│   │   ├── postdown_task
│   │   │   ├── test_playbook.yaml
│   │   │   └── test_var.yaml
│   │   └── post_up
│   │       ├── test_playbook.yaml
│   │       └── test_var.yaml
│   ├── node
│   │   └── testnode
│   │       └── test.js
│   ├── python
│   │   └── do_something
│   │       └── test.py
│   ├── ruby
│   │   └── testruby
│   │       └── test.rb
│   ├── shell
│   │   └── thisisshell.sh
│   └── testtype
│       └── test_something
│           ├── testact.py
│           └── testact.pyc
├── inventories
├── layouts
├── outputs
├── PinFile
├── resources
└── topologies
```

Test: 
- copy pinfile from linchpin/examples/OS_SERVER_PINFILE to workspace
- edit Pinfile <accordingly, if using ruby, node based hooks ruby, node should be installed on host>
- linchpin rise
- linchpin drop
example workspace for testing : https://github.com/samvarankashyap/linchtest

Issues to be resolved:
- Need to add documentation for states and hooks
- Need to add developer documentation for custom_action_manager
- Need to implement hooks with global scope.
Note : refer #96
